packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts

```diff
@@ -0,0 +1,228 @@
+import { ContractWrappers, ContractWrappersError, ForwarderWrapperError, SignedOrder } from '@0x/contract-wrappers';
+import { AbiEncoder, abiUtils, BigNumber, providerUtils } from '@0x/utils';
+import { SupportedProvider, Web3Wrapper, ZeroExProvider } from '@0x/web3-wrapper';
+import { MethodAbi } from 'ethereum-types';
+import * as _ from 'lodash';
+
+import { constants } from '../constants';
+import {
+    AssetSwapQuoterError,
+    CalldataInformation,
+    SmartContractParams,
+    SwapQuote,
+    SwapQuoteConsumer,
+    SwapQuoteConsumerOpts,
+    SwapQuoteExecutionOpts,
+    SwapQuoteGetOutputOpts,
+    SwapQuoteInfo,
+} from '../types';
+import { assert } from '../utils/assert';
+import { utils } from '../utils/utils';
+
+import { assetDataUtils } from '../utils/asset_data_utils';
+
+export interface ForwarderSwapQuoteGetOutputOpts extends SwapQuoteGetOutputOpts {
+    feePercentage: number;
+    feeRecipient: string;
+    ethAmount?: BigNumber;
+}
+
+export interface ForwarderSwapQuoteExecutionOpts extends SwapQuoteExecutionOpts, ForwarderSwapQuoteGetOutputOpts {
+}
+
+const addAffiliateFeeToSwapQuoteInfo = (quoteInfo: SwapQuoteInfo, feePercentage: number): SwapQuoteInfo => {
+    const newQuoteInfo = _.clone(quoteInfo);
+    const affiliateFeeAmount = newQuoteInfo.takerTokenAmount.multipliedBy(feePercentage).integerValue(BigNumber.ROUND_CEIL);
+    const newFeeAmount = newQuoteInfo.feeTakerTokenAmount.plus(affiliateFeeAmount);
+    newQuoteInfo.feeTakerTokenAmount = newFeeAmount;
+    newQuoteInfo.totalTakerTokenAmount = newFeeAmount.plus(newQuoteInfo.takerTokenAmount);
+    return newQuoteInfo;
+};
+
+const addAffiliateFeeToSwapQuote = (quote: SwapQuote, feePercentage: number): SwapQuote => {
+    const newQuote = _.clone(quote);
+    newQuote.bestCaseQuoteInfo = addAffiliateFeeToSwapQuoteInfo(newQuote.bestCaseQuoteInfo, feePercentage);
+    newQuote.worstCaseQuoteInfo = addAffiliateFeeToSwapQuoteInfo(newQuote.worstCaseQuoteInfo, feePercentage);
+    return newQuote;
+};
+
+export class ForwarderSwapQuoteConsumer implements SwapQuoteConsumer {
+
+    public readonly provider: ZeroExProvider;
+    public readonly networkId: number;
+
+    private readonly _contractWrappers: ContractWrappers;
+
+    constructor(
+        supportedProvider: SupportedProvider,
+        options: Partial<SwapQuoteConsumerOpts> = {},
+    ) {
+        const { networkId } = _.merge(
+            {},
+            constants.DEFAULT_ASSET_SWAP_QUOTER_OPTS,
+            options,
+        );
+        const provider = providerUtils.standardizeOrThrow(supportedProvider);
+        assert.isNumber('networkId', networkId);
```

Comment:
Nit: Let's put this up one line. Try to group logic as closely together as possible.